xarray/core/groupby.py

```diff
@@ -395,7 +408,26 @@ def _factorize(self) -> None:
             self.group_indices,
             self.unique_coord,
             self.full_index,
-        ) = self.factorize()
+        ) = self.grouper.factorize(self.group1d)
+
+
+class Grouper(ABC):
+    @property
```

Comment:
Didn't make this abstractmethod since it'll be deleted once the `squeeze` kwarg is completely deprecated.